bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,419 @@
+
+            self.doc_symbols[overridden_symbol] = original_item
+            self.renamed_symbols[symbol_name].append(overridden_symbol)
+
+        elif package_name in PRIORITY_PACKAGES:
+            overridden_symbol = f"{original_item.package}.{symbol_name}"
+            if overridden_symbol in self.doc_symbols:
+                # If there's still a conflict, add the symbol's group in the middle.
+                overridden_symbol = f"{original_item.package}.{original_item.group}.{symbol_name}"
+
+            self.doc_symbols[overridden_symbol] = original_item
+            self.renamed_symbols[symbol_name].append(overridden_symbol)
+
+        # If we can't specially handle the symbol through its group or package,
+        # fall back to prepending its package name to the front.
+        else:
+            new_symbol = f"{package_name}.{symbol_name}"
+            if new_symbol in self.doc_symbols:
+                # If there's still a conflict, add the symbol's group in the middle.
+                new_symbol = f"{package_name}.{group_name}.{symbol_name}"
+            self.renamed_symbols[symbol_name].append(new_symbol)
+            return new_symbol
+
+    async def refresh_inventory(self) -> None:
+        """Refresh internal documentation inventory."""
+        self.refresh_event.clear()
+        log.debug("Refreshing documentation inventory...")
+        self.inventory_scheduler.cancel_all()
+
+        # Clear the old base URLS and doc symbols to ensure
+        # that we start from a fresh local dataset.
+        # Also, reset the cache used for fetching documentation.
+        self.base_urls.clear()
+        self.doc_symbols.clear()
+        self.renamed_symbols.clear()
+        self.scheduled_inventories.clear()
+        await self.item_fetcher.clear()
+
+        # Run all coroutines concurrently - since each of them performs an HTTP
+        # request, this speeds up fetching the inventory data heavily.
+        coros = [
+            self.update_or_reschedule_inventory(
+                package["package"], package["base_url"], package["inventory_url"]
+            ) for package in await self.bot.api_client.get('bot/documentation-links')
+        ]
+        await asyncio.gather(*coros)
+        log.debug("Finished inventory refresh.")
+        self.refresh_event.set()
+
+    async def get_symbol_embed(self, symbol: str) -> Optional[discord.Embed]:
+        """
+        Attempt to scrape and fetch the data for the given `symbol`, and build an embed from its contents.
+
+        If the symbol is known, an Embed with documentation about it is returned.
+
+        First check the DocRedisCache before querying the cog's `BatchParser`.
+        """
+        log.trace(f"Building embed for symbol `{symbol}`")
+        if not self.refresh_event.is_set():
+            log.debug("Waiting for inventories to be refreshed before processing item.")
+            await self.refresh_event.wait()
```

Comment:
Wouldn't that force the gets to wait on each other if we lock them like every command that does a refresh is?
We don't care about the gets running concurrently, but running it while refreshing can lead to a valid symbol being temporarily unavailable

Reply:
Yeah, you're right. A lock isn't the right kind of synchronisation primitive for this since there's no way to make this function avoid locking itself.